### ISSUE TITLE
Error: Index was outside the bounds of the array

### ISSUE THREAD

**Issue by Davilarek** [juanjp600/B3DDecomp#18]
```
Unhandled exception. System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Blitz3DDecomp.DecompilerSteps.Step5.ConvertConstantsToFinalRepresentation.<ProcessStatement>g__map|1_1(Expression expression, Nullable`1 declType) in /home/davil/AUR/B3DDecomp/Blitz3DDecomp/DecompilerSteps/Step5/ConvertConstantsToFinalRepresentation.cs:line 183
   at Blitz3DDecomp.DecompilerSteps.Step5.ConvertConstantsToFinalRepresentation.ProcessStatement(Function function, Int32 statementIndex) in /home/davil/AUR/B3DDecomp/Blitz3DDecomp/DecompilerSteps/Step5/ConvertConstantsToFinalRepresentation.cs:line 326
   at Blitz3DDecomp.DecompilerSteps.Step5.ConvertConstantsToFinalRepresentation.Process(Function function) in /home/davil/AUR/B3DDecomp/Blitz3DDecomp/DecompilerSteps/Step5/ConvertConstantsToFinalRepresentation.cs:line 375
   at Program.Step5(String decompPath, IEnumerable`1 referencedFunctions) in /home/davil/AUR/B3DDecomp/Blitz3DDecomp/Program.cs:line 325
   at Program.Main(String[] args) in /home/davil/AUR/B3DDecomp/Blitz3DDecomp/Program.cs:line 72
```
I would really love to submit a sample where this is happening, but unfortunately as far as I know the game is still copyrighted material (the binary).

**Comment by juanjp600** [comment]
As far as I can tell, the issue here is that a call to a function ended up having less arguments to a function than it needed, which can have many different causes.

Unfortunately, without the binary it'd be nearly impossible for me to figure out what exactly is causing problems here. Could I at least have a name for it so I can try to find it?

**Comment by Davilarek** [comment]
The game is "Zip Zoom Racer".
It's extremely hard to find it nowadays, as it's an abandonware game from 2009 that was sold locally in Russia and Poland.
Someone has posted installer for it on Mediafire.
https://youtu.be/82UmprPCEaI
It's the second link in description of this video.
You may want to try to unpack the installer using some installer unpacker software and then, the main binary is `ZZR.exe`.
Hope this helps.

**Comment by juanjp600** [comment]
Yep, that certainly helps, thanks! At a glance this looks like it's be related to #10, the amount of userlib functions that don't have stdcall symbol decoration seems to be throwing off the parameter guessing step. It could also be the case that this was made with an old version of Blitz3D, which could also break this step. I'll investigate further and hopefully come up with a fix soon.

**Comment by juanjp600** [comment]
Should be fixed in fbd87b9.

The issue was that the decompiler was looking for a second parameter for the `PlayMusic` function. It turns out that [this was added in Third Subdivision's fork of Blitz3D](https://github.com/Saalvage/Blitz3D/commit/48622ccabc4bfe9d281497ff101f855b4a90d565#diff-0934d46969358efca19e836ad3b559f0491b0106aecb25e8d8ce3d645b5974ee).

To fix this problem and many others like it, I made the disassembler look through the data sections of the executable for strings that declare the signature of the built-in functions, which the decompiler then uses instead of the hardcoded ones I extracted from the source code.
